tests/test-dladdr.c

```diff
@@ -0,0 +1,195 @@
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#elif defined(_MSC_VER)
+/* disable warning C4273 inconsistent dll linkage */
+#pragma warning(push)
+#pragma warning(disable: 4273)
+#endif
+/* fix for i586-mingw32msvc-gcc */
+#ifndef WINBASEAPI
+#define WINBASEAPI __declspec(dllimport)
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+WINBASEAPI HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);
+WINBASEAPI int __cdecl atoi(const char *_Str);
+
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+#pragma warning(pop)
+#endif
+#endif
+
+#ifdef _WIN32
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#define PassWithDifferentAddressOnWin PassWithDifferentAddress
+#else
+#define PassWithoutSymbolOnWin Pass
+#define PassWithDifferentAddressOnWin Pass
+#endif
+
+#ifdef DLFCN_WIN32_SHARED
+#define PassWithSymbolOnSharedBuild Pass
+#else
+#define PassWithSymbolOnSharedBuild PassWithoutSymbol
+#endif
+
+#define UNUSED(x) (void)(x)
+
+int main(int argc, char **argv)
+{
+    UNUSED(argv);
+
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    /* null pointer */
+    result = check_dladdr((void*)0, "null" , NoInfo);
+    /* invalid pointer */
+    result |= check_dladdr((void*)0x125, "invalid pointer" , NoInfo);
+    /* -ldl */
+    result |= check_dladdr( ((void*)dladdr), "dladdr" , PassWithSymbolOnSharedBuild );
+    /* -ldl */
+    result |= check_dladdr( (void*)dlopen, "dlopen", PassWithSymbolOnSharedBuild );
+    /* -lglibc */
+    result |= check_dladdr( (void*)atoi, "atoi", Pass );
+    /* test-dladdr */
+    result |= check_dladdr( (void*)main, "main", PassWithoutSymbolOnWin );
+    /* offsets */
+    result |= check_dladdr( (char*)dladdr-6, "dladdr-6", Fail );
+    /* offsets */
+    result |= check_dladdr( (char*)dladdr+6, "dladdr+6", Fail );
```

Comment:
On Linux this test returns:
```
check address 0x7f59c2b10ae6 which has symbol 'dladdr+6' -> passed (filename: /lib/x86_64-linux-gnu/libdl.so.2 base: 0x7f59c2b0f000 symbol name: 'dladdr' addr: 0x7f59c2b10ae0)
```
So it is not failure.

This should rather check that it passes, returns valid `addr` field and corresponding `symbol name` field. `addr` field should be of course different than passed.

Reply:
Returning 'dladdr' in this case requires extending the iat search to a range search.